[RFC] scripts: ci: check_compliance: Add python lint/format check

pdgendt opened the pull request:
## Introduction

As the codebase and contributions keep growing, so does all the python tooling.
This PR is intended to make python scripts more uniform and safe.

### Proposed change

Add [ruff](https://docs.astral.sh/ruff/) as a linter and formatter for python files, and verify (new) files in CI.

## Detailed RFC

Some hightlights:

- `ruff` is a drop-in replacement for many popular python tools, and it is super fast.
- The formatter output is [PEP8](https://peps.python.org/pep-0008/) compliant and produces nearly identical output to [black](https://black.readthedocs.io/en/stable/)
- Can fix issues for you
- ...

Current enabled linter rules:

- [x] [flake8-bugbear](https://docs.astral.sh/ruff/rules/#flake8-bugbear-b)
- [x] [pycodestyle](https://docs.astral.sh/ruff/rules/#pycodestyle-e-w)
- [x] [pyflakes ](https://docs.astral.sh/ruff/rules/#pyflakes-f)
- [x] [isort](https://docs.astral.sh/ruff/rules/#isort-i) (black profile)
- [x] [flake8-simplify](https://docs.astral.sh/ruff/rules/#flake8-simplify-sim)
- [x] [pyupgrade](https://docs.astral.sh/ruff/rules/#pyupgrade-up)

Many more can be added/enabled but this is a solid baseline.

### Concerns

Existing files do not comply, so they are explicitly excluded. Ideally these are gradually converted to comply.

```sh
$ ruff check
Found 3708 errors.

$ ruff format --check
379 files would be reformatted, 34 files already formatted
```

Formatting is obviously opinionated, but I do think for python we should follow the popular kids, which are `black` and `ruff`.

pdgendt committed:
scripts: ci: check_compliance: Add python lint/format check

Add a compliance test using ruff, for both linting and formatting of
newly added python files.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
scripts: ci: check_compliance: Add support for end line and column

Reporting or annotating issues can be done on a range rather than a
single line.

Add support for end line and end column.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt commented:
> Why add exclusions? If someone adds a new file e.g. a runner from an external source, they would have to fix the file, but if an existing runner which does not comply is excluded, that can be updated forever going forwards without having to comply? Or is the idea that new exclusions would be added, in which case what would be the point of having the check if everything is excluded?

This is up for debate off course, but the idea is to force new files to comply, while we gradually convert the existing ones. It could be brought up during a review to update a file you are touching to comply.

nordicjm commented:
> > Why add exclusions? If someone adds a new file e.g. a runner from an external source, they would have to fix the file, but if an existing runner which does not comply is excluded, that can be updated forever going forwards without having to comply? Or is the idea that new exclusions would be added, in which case what would be the point of having the check if everything is excluded?
> 
> This is up for debate of course, but the idea is to force new files to comply, while we gradually convert the existing ones. It could be brought up during a review to update a file you are touching to comply.

This is something I asked about before because one of the files added from an external source, a uf2 one if I remember right, had some oddities about it, and someone said the file shouldn't be fixed because whilst it was being added to zephyr, it was not a zephyr file, and you then lose the ability to diff etc.

pdgendt commented:
> This is something I asked about before because one of the files added from an external source, a uf2 one if I remember right, had some oddities about it, and someone said the file shouldn't be fixed because whilst it was being added to zephyr, it was not a zephyr file, and you then lose the ability to diff etc.

I think this can be an exception, but that's what the exclude lists are for.

hakehuang commented:
> * Can fix issues for you

can we use ruff to fix the issues in this PR?

pdgendt commented:
> can we use ruff to fix the issues in this PR?

Kind of chicken and egg, the `check_compliance.py` file itself is currently excluded because already existing issues.

```shell
$ ruff format --check scripts/ci/check_compliance.py
Would reformat: scripts/ci/check_compliance.py
1 file would be reformatted

$ ruff check scripts/ci/check_compliance.py
Found 25 errors.
```

pdgendt committed:
scripts: Add helper scripts for ruff baseline excludes

Add simple scripts to convert ruff check and ruff format output to
toml exclude sections.

These sections can be used to ignore baseline violations for an existing
codebase.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
scripts: ci: Add ruff configuration files

Add a baseline toml file for current rule violations, and a default
configuration file.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

hakehuang commented:
> Kind of chicken and egg, the `check_compliance.py` file itself is currently excluded because of already existing issues.

so if this PR merged, can we then use ruff to auto fix issues? if not, this could be a heavy loading

pdgendt commented:
> so if this PR merged, can we then use ruff to auto fix issues? if not, this could be a heavy loading

Yes! For the current selected rules:

```shell
$ ruff check
...
Found 3713 errors.
[*] 1972 fixable with the `--fix` option (506 hidden fixes can be enabled with the `--unsafe-fixes` option).
```

But I'd suggest doing fixes in separate PRs to make it easier to digest for reviewing, and it's best to double check automatic fixes.

Formatting can be done too, but this would result in a lot of git churn.

fabiobaltieri commented:
Seems like a good idea, curious why did you pick ruff instead of black? (I know pretty much nothing about these)

kartben commented:
> Seems like a good idea, curious why did you pick ruff instead of black? (I know pretty much nothing about these)

I think black is just a formatter while ruff does both linting & formatting?

pdgendt commented:
> Seems like a good idea, curious why did you pick ruff instead of black? (I know pretty much nothing about these)

It's all packed into one tool, so it should never contradict itself, while using different linter and formatter tools might need exceptions.
It's also super fast, and pleasant to work with. It can also be used as an LSP to get feedback directly in your IDE.

pdgendt commented:
Rebased on `main`

pdgendt commented:
Update:

- Ignore `SIM108` to allow if-else blocks instead of forcing ternary operator

fabiobaltieri commented:
@pdgendt where do you want to go from here? It's titled and tagged as RFC  but arguably good to go. Maybe arch wg? (though it's not really an architectural change but seems like any change with wide impact goes there these days)

pdgendt commented:
I was thinking the same thing, wanted to bring this up during some workgroup meeting. Could be for arch or process? Maybe we need a tooling working group?

CC @carlescufi @nashif any suggestions?

fabiobaltieri commented:
I think process would be the "correct-er" one but if you want visibility arch is the one to go, I suggest bringing it up there.

pdgendt commented:
> are we going to keep pylint?

There's no conflict between these two, there is [some overlap](https://docs.astral.sh/ruff/faq/#how-does-ruffs-linter-compare-to-pylint) but I haven't added any of the `PL` rules in this PR.

We might favor enabling the `PL` rules from `ruff` and replace `pylint` with an actual type checker. But that's outside of the scope of this PR.

nashif commented:
> There's no conflict between these two, there is [some overlap](https://docs.astral.sh/ruff/faq/#how-does-ruffs-linter-compare-to-pylint) but I haven't added any of the `PL` rules in this PR.
> 
> We might favor enabling the `PL` rules from `ruff` and replace `pylint` with an actual type checker. But that's outside of the scope of this PR.

but it is suboptimal if we have two linters doing python, at leasr from a developer experience. For example, if there is a violation that is detected by both, will we get two warnings?

pdgendt commented:
> but it is suboptimal if we have two linters doing python, at leasr from a developer experience. For example, if there is a violation that is detected by both, will we get two warnings?

Presumably, yes. But I haven't looked into the overlaps of the current configurations, this can be optimized on the go.

As you can see from the excludes file, there are a lot of linter issues that `pylint` hasn't reported on.

fabiobaltieri commented:
I think that before doing this (or at least as part of this) we should at least batch fix the files that are obviously not imported and meant to track an external project, I guess anything under arch, doc, samples, script, tests...

pdgendt commented:
> I think that before doing this (or at least as part of this) we should at least batch fix the files that are obviously not imported and meant to track an external project, I guess anything under arch, doc, samples, script, tests...

I was going to do it the other way around, this PR is to prevent any new issues from this point forward.

fabiobaltieri commented:
> I was going to do it the other way around, this PR is to prevent any new issues from this point forward.

Yup that makes sense.

fabiobaltieri commented:
- summarize the issue: linter+formatter check for any new python file
- @nashif: asks how this behaves with the current files (out of compliance), current files/exclusions woul
d not trigger
- @pdgendt: most of these can be fixed esily, will do some folowup to look into this, the tool has an option
 to change the file inline and also do only "safe" changes
- no conerns on trying this for new files

andyross commented:
FYI: this isn't running only for new files, as it seems like people were expecting.  It's now running on all changes to **any** python file, to, um hilarious effect, see #81154

kartben commented:
> FYI: this isn't running only for new files, as it seems like people were expecting.  It's now running on all changes to **any** python file, to, um hilarious effect, see #81154

As noted in the linked issue, you moved existing files so updating the ruff-excludes entries accordingly should bring you back on track.
